### ISSUE TITLE
[DependencyInjection] Access Denied while trying to use INamedBucketProvider

### ISSUE THREAD

**Issue by ZackStone** [brantburnett/Couchbase.Extensions#14]
Hi!

I'm trying to use the interface `INamedBucketProvider`, but I'm getting the "Access Denied" error.
When I try to connect using just the `IBucketProvider`, it works perfectly.

Here are some important parts of my code:

```c#
public interface ICustomerBucketProvider : INamedBucketProvider
{
    // nothing goes in here!
}
```
```c#
public static class IServiceCollectionExtension
{
    public static IServiceCollection AddCouchbaseConnector(this IServiceCollection services)
    {
        services.AddCouchbase(opt => 
        {
            opt.ConnectionString = "couchbase://localhost";
            opt.UserName = "Administrator";
            opt.Password = "asd123";
            opt.Buckets = new[] { "customers" };
        });

        services.AddCouchbaseBucket<ICustomerBucketProvider>("customers");

        return services;
    }
}
```
```c#
public class NativeInjectorBootStrapper
{
    public static void RegisterServices(IServiceCollection services)
    {
        // ...
        // ...
        // ...
        
        services.AddCouchbaseConnector();
    }
}
```

This works:
```c#
private readonly IBucketProvider _bucketProvider;

public CustomerEventHandler(IBucketProvider bucketProvider)
{
    _bucketProvider = bucketProvider;
}

private async Task<ICouchbaseCollection> GetCouchbaseCollection()
{
    var bucket = await _bucketProvider.GetBucketAsync("customers");
    var collection = bucket.DefaultCollection();
    return collection;
}
```

And this doesn't work:
```c#
private readonly ICustomerBucketProvider _customerBucketProvider;

public CustomerEventHandler(ICustomerBucketProvider customerBucketProvider)
{
    _customerBucketProvider = customerBucketProvider;
}

private async Task<ICouchbaseCollection> GetCouchbaseCollection()
{
    var bucket = await _customerBucketProvider.GetBucketAsync();
    var collection = bucket.DefaultCollection();
    return collection;
}
```

Am I doing something wrong?

**Comment by brantburnett** [comment]
The most likely problem may be related to Castle Dynamic Proxies. That is used to dynamically create a class that implements the provided ICustomerBucketProvider interface at runtime.

My first thought would have been that the interface was marked internal, and you needed an InternalsVisibleTo attribute on your assembly. But that doesn't seem to be the case.

Are you running in a locked-down IIS environment on the full .NET framework? It might be blocking dynamic proxy generation or something like that.

**Comment by ZackStone** [comment]
I'm using a MVC web app on .NET Core 3.1

**Comment by brantburnett** [comment]
Interesting. I know we're using this without problems in .NET Core 3.1. Can you provide more details on the exception like the stack trace?

**Comment by ZackStone** [comment]
- Exception details:
```
System.TypeLoadException
  HResult=0x80131522
  Message=Access is denied: 'Couchbase.Extensions.DependencyInjection.Internal.NamedBucketProvider'.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>
```
- ($exception).StackTrace:
```
   at System.Reflection.Emit.TypeBuilder.CreateTypeNoLock()
   at System.Reflection.Emit.TypeBuilder.CreateTypeInfo()
   at Couchbase.Extensions.DependencyInjection.Internal.NamedBucketProxyGenerator.CreateProxyType(Type interfaceType)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Couchbase.Extensions.DependencyInjection.Internal.NamedBucketProxyGenerator.GetProxy[T](IBucketProvider bucketProvider, String bucketName)
   at Couchbase.Extensions.DependencyInjection.ServiceCollectionExtensions.<>c__DisplayClass2_0`1.<AddCouchbaseBucket>b__0(IServiceProvider serviceProvider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitIEnumerable(IEnumerableCallSite enumerableCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at MediatR.ServiceFactoryExtensions.GetInstances[T](ServiceFactory factory)
   at MediatR.Internal.NotificationHandlerWrapperImpl`1.Handle(INotification notification, CancellationToken cancellationToken, ServiceFactory serviceFactory, Func`4 publish)
   at MediatR.Mediator.PublishNotification(INotification notification, CancellationToken cancellationToken)
   at MediatR.Mediator.Publish[TNotification](TNotification notification, CancellationToken cancellationToken)
   at Equinox.Infra.CrossCutting.Bus.InMemoryBus.RaiseEvent[T](T event) in C:\Users\BSI99491\Downloads\EquinoxProject-master\src\Equinox.Infra.CrossCutting.Bus\InMemoryBus.cs:line 30
```
- ($exception).TargetSite:
```
    {System.Reflection.TypeInfo CreateTypeNoLock()}	System.Reflection.MethodBase {System.Reflection.RuntimeMethodInfo}
```

**Comment by brantburnett** [comment]
Interesting, this does seem to be related to public vs internal in some way. Two things to try:

1. Make sure your interface is public
2. Add this attribute to your assembly:

[assembly: InternalsVisibleTo(InternalsVisible.ToDynamicProxyGenAssembly2)]

**Comment by ZackStone** [comment]
Same error... I decided to implement my own structure to handle buckets and collections:

```c#
public interface ICouchbaseBucketBase
{
    public ICouchbaseCollection DefaultCollection { get; }
}

public class CouchbaseBucketBase : ICouchbaseBucketBase
{
    public ICouchbaseCollection DefaultCollection { get; private set; }

    protected readonly IBucket _bucket;

    public CouchbaseBucketBase(IBucketProvider bucketProvider, string bucketName)
    {
        _bucket = bucketProvider.GetBucketAsync(bucketName).GetAwaiter().GetResult();

        DefaultCollection = GetCollection();
    }

    protected ICouchbaseCollection GetCollection(string collectionName = null) =>
        !string.IsNullOrWhiteSpace(collectionName) ?
        _bucket.Collection(collectionName) : _bucket.DefaultCollection();
}
```

```c#
public interface ICustomerBucket : ICouchbaseBucketBase { }

public class CustomerBucket : CouchbaseBucketBase, ICustomerBucket
{
    //public ICouchbaseCollection OtherNonDefaultCollection { get; private set; }

    public CustomerBucket(IBucketProvider bucketProvider) : base(bucketProvider, "customers")
    {
        //OtherNonDefaultCollection = GetCollection("OtherNonDefault");
    }
}
```

```c#
public static class IServiceCollectionExtension
{
    public static IServiceCollection AddCouchbaseConnector(this IServiceCollection services)
    {
        services.AddCouchbase(opt =>
        {
            opt.ConnectionString = "couchbase://localhost";
            opt.UserName = "Administrator";
            opt.Password = "asd123";
            opt.Buckets = new[] { "customers" };
        });

        services.AddTransient<ICustomerBucket, CustomerBucket>();

        return services;
    }
}
```

**Comment by mtinnes** [comment]
Same issue here, using .NET core 3.1.4. Very frustrating as it seems .NET is not considered a top tier supported language by Couchbase.

**Comment by brantburnett** [comment]
I believe we've tracked down the problem and we're working on a fix. Some changes to support CI delivery of the NuGet package for strongly-named assemblies broke some required InternalsVisibleTo attributes. But only the NuGet distro was affected, all of the dev machine tests were passing.

https://issues.couchbase.com/browse/NCBC-2558

**Comment by brantburnett** [comment]
@ZackStone @mtinnes This issue should be fixed in the upcoming 3.0.2 release of the SDK. Thanks for letting us know about it!
